File: CSharp/Legacy89DiskKit.Infrastructure/DiskImage/Container/RawDiskContainer.cs
using Legacy89DiskKit.Domain.DiskImage.Interface.Container;
using Legacy89DiskKit.Domain.DiskImage.Model;
using Legacy89DiskKit.Domain.DiskImage.Exception;

namespace Legacy89DiskKit.Infrastructure.DiskImage.Container;

public class RawDiskContainer : IDiskContainer
{
    private readonly byte[] _diskData;
    private readonly string _filePath;
    private readonly bool _readOnly;
    private bool _hasChanges = false;
    
    private readonly int _cylinders;
    private readonly int _sides;
    private readonly int _sectorsPerTrack;
    private readonly int _bytesPerSector;
    private readonly DiskType _diskType;

    public RawDiskContainer(string filePath, bool readOnly = true)
        : this(LoadDiskImage(filePath), filePath, readOnly)
    {
    }

    /// <summary>
    /// Initializes a raw disk container from an in-memory disk image.
    /// </summary>
    public RawDiskContainer(byte[] diskData, bool readOnly = true, string filePath = "")
        : this(diskData ?? throw new ArgumentNullException(nameof(diskData)), filePath, readOnly)
    {
    }

    private RawDiskContainer(byte[] diskData, string filePath, bool readOnly)
    {
        _filePath = filePath ?? "";
        _readOnly = readOnly;
        _diskData = (byte[])diskData.Clone();
        
        (_cylinders, _sides, _sectorsPerTrack, _bytesPerSector, _diskType) = DetectGeometry(_diskData.Length);
    }

    private static byte[] LoadDiskImage(string filePath)
    {
        if (filePath is null)
        {
            throw new ArgumentNullException(nameof(filePath));
        }

        if (!File.Exists(filePath))
        {
            throw new FileNotFoundException($"Disk image file not found: {filePath}");
        }

        return File.ReadAllBytes(filePath);
    }

    private static (int c, int h, int spt, int bps, DiskType type) DetectGeometry(long size)
    {
        return size switch
        {
            327680 => (40, 2, 16, 256, DiskType.TwoD),    // X1/PC88 2D
            655360 => (80, 2, 16, 256, DiskType.TwoDD),   // X1/PC88 2DD (256B)
            737280 => (80, 2, 9, 512, DiskType.TwoDD),    // MSX/PC 2DD (512B)
            1261568 => (77, 2, 8, 1024, DiskType.TwoHD),  // PC-98 2HD
            1474560 => (80, 2, 18, 512, DiskType.TwoHD),  // PC 2HD
            _ => (40, 2, 16, 256, DiskType.TwoD)           // Default to 2D but warn?
        };
    }

    public static RawDiskContainer CreateNew(string filePath, DiskType type, int? sectorsPerTrack = null, ushort? sectorSize = null)
    {
        var data = CreateEmptyDiskData(type);
        File.WriteAllBytes(filePath, data);
        return new RawDiskContainer(data, false, filePath);
    }

    /// <summary>
    /// Creates a new in-memory raw disk container.
    /// </summary>
    public static RawDiskContainer CreateNewInMemory(DiskType type, int? sectorsPerTrack = null, ushort? sectorSize = null)
    {
        return new RawDiskContainer(CreateEmptyDiskData(type), false);
    }

    private static byte[] CreateEmptyDiskData(DiskType type)
    {
        int size = type switch
        {
            DiskType.TwoD => 327680,
            DiskType.TwoDD => 720 * 1024,
            DiskType.TwoHD => 1440 * 1024,
            _ => 327680
        };

        return new byte[size];
    }

    public byte[] ReadSector(int cylinder, int head, int sector) => ReadSector(cylinder, head, sector, false);

    public byte[] ReadSector(int cylinder, int head, int sector, bool allowCorrupted)
    {
        ValidateAddress(cylinder, head, sector);
        int offset = CalculateOffset(cylinder, head, sector);
        byte[] sectorData = new byte[_bytesPerSector];
        Array.Copy(_diskData, offset, sectorData, 0, _bytesPerSector);
        return sectorData;
    }

    public void WriteSector(int cylinder, int head, int sector, byte[] data)
    {
        if (_readOnly) throw new DiskImageException("Disk image is read-only");
        ValidateAddress(cylinder, head, sector);
        if (data.Length != _bytesPerSector)
            throw new ArgumentException($"Sector size must be {_bytesPerSector} bytes");

        int offset = CalculateOffset(cylinder, head, sector);
        Array.Copy(data, 0, _diskData, offset, _bytesPerSector);
        _hasChanges = true;
    }

    public bool SectorExists(int cylinder, int head, int sector)
    {
        return cylinder >= 0 && cylinder < _cylinders &&
               head >= 0 && head < _sides &&
               sector >= 1 && sector <= _sectorsPerTrack;
    }

    public IEnumerable<SectorInfo> GetAllSectors()
    {
        for (int c = 0; c < _cylinders; c++)
        {
            for (int h = 0; h < _sides; h++)
            {
                for (int s = 1; s <= _sectorsPerTrack; s++)
                {
                    yield return new SectorInfo(c, h, s, _bytesPerSector, false, false);
                }
            }
        }
    }

    public void Save()
    {
        if (_readOnly) throw new DiskImageException("Cannot save read-only disk image");
        if (string.IsNullOrEmpty(_filePath)) throw new DiskImageException("Cannot save a disk image without a file path");
        File.WriteAllBytes(_filePath, _diskData);
    }

    public void SaveAs(string filePath) => File.WriteAllBytes(filePath, _diskData);

    /// <summary>
    /// Returns the current disk image bytes.
    /// </summary>
    public byte[] ToImageData() => (byte[])_diskData.Clone();

    private int CalculateOffset(int cylinder, int head, int sector)
    {
        return ((cylinder * _sides + head) * _sectorsPerTrack + (sector - 1)) * _bytesPerSector;
    }

    private void ValidateAddress(int cylinder, int head, int sector)
    {
        if (!SectorExists(cylinder, head, sector))
            throw new ArgumentOutOfRangeException($"Invalid sector address: C:{cylinder} H:{head} S:{sector}");
    }

    public string FilePath => _filePath;
    public bool IsReadOnly => _readOnly;
    public DiskType DiskType => _diskType;

    public void Dispose() 
    {
        if (_hasChanges && !_readOnly && !string.IsNullOrEmpty(_filePath))
        {
            try { Save(); } catch { }
        }
    }
}

File: Cpp/Legacy89DiskKit.Cpp/src/infrastructure/disk_image/raw_disk_container.cpp
#include "legacy89diskkit/cpp/infrastructure/disk_image/raw_disk_container.hpp"

#include "legacy89diskkit/cpp/infrastructure/disk_image/raw_buffer_loader.hpp"
#include "legacy89diskkit/cpp/raw_disk_geometry.hpp"
#include "legacy89diskkit/cpp/disk_image_types.hpp"

namespace legacy89diskkit::cpp
{
Result<RawDiskContainer> RawDiskContainer::OpenFromBuffer(const std::span<const std::uint8_t> image_data, const bool read_only)
{
    const auto layout = RawBufferLoader::Load(image_data);
    if (!layout.ok())
    {
        return Result<RawDiskContainer>::Failure(layout.status().code, layout.status().message);
    }

    auto metadata = layout.value().metadata;
    metadata.is_write_protected = metadata.is_write_protected || read_only;
    return Result<RawDiskContainer>::Success(
        RawDiskContainer(std::vector<std::uint8_t>(image_data.begin(), image_data.end()), std::move(metadata), read_only, {}));
}

Result<RawDiskContainer> RawDiskContainer::CreateNew(DiskType type)
{
    std::size_t size = 0;
    int cylinders = 40;
    int heads = 2;
    int sectors_per_track = 16;
    int sector_size = 256;

    switch (type)
    {
    case DiskType::TwoD:
        size = 327680;
        cylinders = 40;
        heads = 2;
        sectors_per_track = 16;
        sector_size = 256;
        break;
    case DiskType::TwoDD:
        size = 720 * 1024;
        cylinders = 80;
        heads = 2;
        sectors_per_track = 9;
        sector_size = 512;
        break;
    case DiskType::TwoHD:
        size = 1440 * 1024;
        cylinders = 80;
        heads = 2;
        sectors_per_track = 18;
        sector_size = 512;
        break;
    default:
        return Result<RawDiskContainer>::Failure(StatusCode::InvalidArgument, "Unsupported disk type for raw creation.");
    }

    DiskContainerMetadata metadata{};
    metadata.disk_type = type;
    metadata.geometry = DiskGeometryInfo{cylinders, heads, sectors_per_track, sector_size};
    metadata.is_write_protected = false;
    metadata.image_format = "raw-sector-container";

    return Result<RawDiskContainer>::Success(
        RawDiskContainer(std::vector<std::uint8_t>(size, 0x00), std::move(metadata), false, {}));
}

RawDiskContainer::RawDiskContainer(
    std::vector<std::uint8_t> image_data,
    DiskContainerMetadata metadata,
    const bool read_only,
    std::string file_path)
    : image_data_(std::move(image_data)),
      metadata_(std::move(metadata)),
      read_only_(read_only),
      file_path_(std::move(file_path))
{
}

const std::string& RawDiskContainer::FilePath() const
{
    return file_path_;
}

bool RawDiskContainer::IsReadOnly() const
{
    return read_only_;
}

DiskType RawDiskContainer::DiskTypeValue() const
{
    return metadata_.disk_type;
}

DiskContainerMetadata RawDiskContainer::GetMetadata() const
{
    return metadata_;
}

Result<std::vector<std::uint8_t>> RawDiskContainer::ReadSector(const int cylinder, const int head, const int sector) const
{
    RawSectorAddressCalculator calculator(RawDiskGeometry{
        metadata_.geometry.cylinders,
        metadata_.geometry.heads,
        metadata_.geometry.sectors_per_track,
        metadata_.geometry.bytes_per_sector,
        metadata_.disk_type});

    const auto offset = calculator.CalculateOffset(cylinder, head, sector);
    if (!offset.ok())
    {
        return Result<std::vector<std::uint8_t>>::Failure(offset.status().code, offset.status().message);
    }

    const auto size = static_cast<std::size_t>(metadata_.geometry.bytes_per_sector);
    return Result<std::vector<std::uint8_t>>::Success(
        std::vector<std::uint8_t>(
            image_data_.begin() + offset.value(),
            image_data_.begin() + offset.value() + static_cast<std::ptrdiff_t>(size)));
}

Status RawDiskContainer::WriteSector(const int cylinder, const int head, const int sector, const std::span<const std::uint8_t> data)
{
    if (read_only_)
    {
        return {StatusCode::InvalidArgument, "Raw disk container is read-only."};
    }

    if (data.size() > static_cast<std::size_t>(metadata_.geometry.bytes_per_sector))
    {
        return {StatusCode::InvalidArgument, "Data size exceeds sector size."};
    }

    RawSectorAddressCalculator calculator(RawDiskGeometry{
        metadata_.geometry.cylinders,
        metadata_.geometry.heads,
        metadata_.geometry.sectors_per_track,
        metadata_.geometry.bytes_per_sector,
        metadata_.disk_type});

    const auto offset = calculator.CalculateOffset(cylinder, head, sector);
    if (!offset.ok())
    {
        return offset.status();
    }

    std::copy(data.begin(), data.end(), image_data_.begin() + offset.value());
    dirty_ = true;
    return Status::OkStatus();
}

bool RawDiskContainer::SectorExists(const int cylinder, const int head, const int sector) const
{
    RawSectorAddressCalculator calculator(RawDiskGeometry{
        metadata_.geometry.cylinders,
        metadata_.geometry.heads,
        metadata_.geometry.sectors_per_track,
        metadata_.geometry.bytes_per_sector,
        metadata_.disk_type});
    return calculator.SectorExists(cylinder, head, sector);
}

std::vector<SectorInfo> RawDiskContainer::GetAllSectors() const
{
    std::vector<SectorInfo> sectors;
    sectors.reserve(static_cast<std::size_t>(metadata_.geometry.cylinders) *
        static_cast<std::size_t>(metadata_.geometry.heads) *
        static_cast<std::size_t>(metadata_.geometry.sectors_per_track));

    for (int cylinder = 0; cylinder < metadata_.geometry.cylinders; ++cylinder)
    {
        for (int head = 0; head < metadata_.geometry.heads; ++head)
        {
            for (int sector = 1; sector <= metadata_.geometry.sectors_per_track; ++sector)
            {
                sectors.push_back(SectorInfo{cylinder, head, sector, metadata_.geometry.bytes_per_sector, false, false});
            }
        }
    }

    return sectors;
}

bool RawDiskContainer::HasChanges() const
{
    return dirty_;
}

void RawDiskContainer::ResetChanges()
{
    dirty_ = false;
}

std::vector<std::uint8_t> RawDiskContainer::ToImageData() const
{
    return image_data_;
}
}
